tests/LexerTest.elm

```diff
@@ -0,0 +1,1086 @@
+module LexerTest exposing (..)
+
+import Dict
+import Elm.AST.Frontend as Frontend
+import Elm.AST.Frontend.Unwrapped exposing (Expr(..), Pattern(..))
+import Elm.Data.Declaration as Declaration exposing (DeclarationBody)
+import Elm.Data.Exposing exposing (ExposedItem(..), Exposing(..))
+import Elm.Data.Located as Located
+import Elm.Data.Module exposing (ModuleType(..))
+import Elm.Data.Qualifiedness exposing (PossiblyQualified(..))
+import Elm.Data.Type.Concrete as ConcreteType exposing (ConcreteType)
+import Elm.Data.TypeAnnotation exposing (TypeAnnotation)
+import Expect exposing (Expectation)
+import OurExtras.String as String
+import Parser.Advanced as P
+import Stage.Parse.Lexer exposing (..)
+import String.Extra as String
+import Test exposing (Test, describe, test)
+
+
+runTest ( description, input ) =
+    test description <|
+        \() ->
+            input
+                -- |> P.run ((located (Stage.Parse.Lexer.literalParser
+                --         |> P.map (\( ty, literalBody ) -> Literal ty literalBody  )))
+                --           |> P.map List.singleton)
+                |> P.run Stage.Parse.Lexer.parser
+                |> Expect.all
+                    [ Result.map (List.map (Located.unwrap >> toString) >> String.join "")
+                        >> Expect.equal (Ok input)
+                    , Result.map
+                        (List.filterMap
+                            (\item ->
+                                case Located.unwrap item of
+                                    Invalid s ->
+                                        Just s
+
+                                    _ ->
+                                        Nothing
+                            )
+                        )
+                        >> Expect.equal (Ok [])
+                    ]
+
+
+moduleDeclaration : Test
+moduleDeclaration =
+    describe "Stage.Parse.Parser.moduleDeclaration"
+        [ describe "general"
+            (List.map runTest
+                [ ( "works with simple module name"
+                  , "module Foo exposing (..)"
+                    -- , [ Token "module"
+                    --   , Whitespace 1
+                    --   , Token "Foo"
+                    --   , Whitespace 1
+                    --   , Token "exposing"
+                    --   , Whitespace 1
+                    --   , Sigil (Bracket Round Open)
+                    --   , Sigil DoubleDot
+                    --   , Sigil (Bracket Round Close)
+                    --   ]
+                  )
+                , ( "works with nested module name"
+                  , "module Foo.Bar exposing (..)"
+                  )
```

Comment:
I would love these tests to not only check that there are no invalid items and that the roundtrip from/to String went OK, but show what the lexed items are... for learning purposes.

Eg. if I saw

```elm
( "works with nested module name"
, "module Foo.Bar exposing (..)"
, [Token "module", ...] -- purposefully stripped of location info, to keep signal/noise ratio up
)
```

that would help me understand what do the results of this lexing process look like.

Kinda like you did in the test just above this one :)

Reply:
Yeah, I went for the quickest way to make the Parser tests work for my Lexer (which was just to remove the bits that required any meaning change).

To complete your example:

```elm
( "works with nested module name"
, "module Foo.Bar exposing (..)"
, [Token "module", Whitespace 1,  Token "Foo", Sigil FullStop, Token "Bar", Whitespace 1,  Token "exposing", Whitespace 1, Sigil (Bracket Round Open), Sigil DoubleDot, Sigil (Bracket Round Close) ]"
)
```

or something along these lines. I.E very long! Unless I can find some way of automating I do not think I can face typing that out for all 100 or so tests in the LexerTests.elm file!

Reply:
Fair enough! The tests were previously written incrementally so the cost was ... amortized :) But right now it's not a huge priority, so I agree with you - let's not add those examples right now.